diff --git a/cpp/src/arrow/io/interfaces.cc b/cpp/src/arrow/io/interfaces.cc
index d052c0168..72b44aa0a 100644
--- a/cpp/src/arrow/io/interfaces.cc
+++ b/cpp/src/arrow/io/interfaces.cc
@@ -49,7 +49,7 @@ namespace io {
 static IOContext g_default_io_context{};
 
 IOContext::IOContext(MemoryPool* pool, StopToken stop_token)
-    : IOContext(pool, internal::GetIOThreadPool(), std::move(stop_token)) {}
+    : IOContext(pool, nullptr, std::move(stop_token)) {}
 
 const IOContext& default_io_context() { return g_default_io_context; }
 
diff --git a/cpp/src/parquet/arrow/reader.cc b/cpp/src/parquet/arrow/reader.cc
index 016ceacb0..c280f700e 100644
--- a/cpp/src/parquet/arrow/reader.cc
+++ b/cpp/src/parquet/arrow/reader.cc
@@ -887,15 +887,6 @@ Status FileReaderImpl::GetRecordBatchReader(const std::vector<int>& row_groups,
                                             std::unique_ptr<RecordBatchReader>* out) {
   RETURN_NOT_OK(BoundsCheck(row_groups, column_indices));
 
-  if (reader_properties_.pre_buffer()) {
-    // PARQUET-1698/PARQUET-1820: pre-buffer row groups/column chunks if enabled
-    BEGIN_PARQUET_CATCH_EXCEPTIONS
-    ARROW_UNUSED(reader_->PreBuffer(row_groups, column_indices,
-                                    reader_properties_.io_context(),
-                                    reader_properties_.cache_options()));
-    END_PARQUET_CATCH_EXCEPTIONS
-  }
-
   std::vector<std::shared_ptr<ColumnReaderImpl>> readers;
   std::shared_ptr<::arrow::Schema> batch_schema;
   RETURN_NOT_OK(GetFieldReaders(column_indices, row_groups, &readers, &batch_schema));
@@ -987,15 +978,6 @@ Status FileReaderImpl::ReadRowGroups(const std::vector<int>& row_groups,
                                      std::shared_ptr<Table>* out) {
   RETURN_NOT_OK(BoundsCheck(row_groups, column_indices));
 
-  // PARQUET-1698/PARQUET-1820: pre-buffer row groups/column chunks if enabled
-  if (reader_properties_.pre_buffer()) {
-    BEGIN_PARQUET_CATCH_EXCEPTIONS
-    ARROW_UNUSED(parquet_reader()->PreBuffer(row_groups, column_indices,
-                                             reader_properties_.io_context(),
-                                             reader_properties_.cache_options()));
-    END_PARQUET_CATCH_EXCEPTIONS
-  }
-
   std::vector<std::shared_ptr<ColumnReaderImpl>> readers;
   std::shared_ptr<::arrow::Schema> result_schema;
   RETURN_NOT_OK(GetFieldReaders(column_indices, row_groups, &readers, &result_schema));
diff --git a/cpp/src/parquet/properties.h b/cpp/src/parquet/properties.h
index 13ddc78cf..26fee95a0 100644
--- a/cpp/src/parquet/properties.h
+++ b/cpp/src/parquet/properties.h
@@ -573,9 +573,7 @@ class PARQUET_EXPORT ArrowReaderProperties {
   explicit ArrowReaderProperties(bool use_threads = kArrowDefaultUseThreads)
       : use_threads_(use_threads),
         read_dict_indices_(),
-        batch_size_(kArrowDefaultBatchSize),
-        pre_buffer_(false),
-        cache_options_(::arrow::io::CacheOptions::Defaults()) {}
+        batch_size_(kArrowDefaultBatchSize) {}
 
   void set_use_threads(bool use_threads) { use_threads_ = use_threads; }
 
@@ -600,33 +598,10 @@ class PARQUET_EXPORT ArrowReaderProperties {
 
   int64_t batch_size() const { return batch_size_; }
 
-  /// Enable read coalescing.
-  ///
-  /// When enabled, the Arrow reader will pre-buffer necessary regions
-  /// of the file in-memory. This is intended to improve performance on
-  /// high-latency filesystems (e.g. Amazon S3).
-  void set_pre_buffer(bool pre_buffer) { pre_buffer_ = pre_buffer; }
-
-  bool pre_buffer() const { return pre_buffer_; }
-
-  /// Set options for read coalescing. This can be used to tune the
-  /// implementation for characteristics of different filesystems.
-  void set_cache_options(::arrow::io::CacheOptions options) { cache_options_ = options; }
-
-  ::arrow::io::CacheOptions cache_options() const { return cache_options_; }
-
-  /// Set execution context for read coalescing.
-  void set_io_context(const ::arrow::io::IOContext& ctx) { io_context_ = ctx; }
-
-  const ::arrow::io::IOContext& io_context() const { return io_context_; }
-
  private:
   bool use_threads_;
   std::unordered_set<int> read_dict_indices_;
   int64_t batch_size_;
-  bool pre_buffer_;
-  ::arrow::io::IOContext io_context_;
-  ::arrow::io::CacheOptions cache_options_;
 };
 
 /// EXPERIMENTAL: Constructs the default ArrowReaderProperties
